src/ucp/core/ucp_ep_vfs.c

```diff
@@ -0,0 +1,153 @@
+         {"remote", UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR,
+          ucs_offsetof(ucp_ep_attr_t, remote_sockaddr)}};
+
+static const char *ucp_err_handling_mode_names[] = {
+    [UCP_ERR_HANDLING_MODE_NONE] = "none",
+    [UCP_ERR_HANDLING_MODE_PEER] = "peer"
+};
+
+static void ucp_ep_vfs_read_peer_name(void *obj, ucs_string_buffer_t *strb,
+                                      void *arg_ptr, uint64_t arg_u64)
+{
+    ucp_ep_h ep = obj;
+
+    ucs_string_buffer_appendf(strb, "%s\n", ucp_ep_peer_name(ep));
+}
+
+static ucs_status_t ucp_ep_vfs_query_sockaddr(ucp_ep_h ep, ucp_ep_attr_t *attr,
+                                              uint64_t field_mask,
+                                              ucs_string_buffer_t *strb)
+{
+    ucs_status_t status;
+
+    attr->field_mask = field_mask;
+    status           = ucp_ep_query_sockaddr(ep, attr);
+    if (status != UCS_OK) {
+        ucs_string_buffer_appendf(strb, "<not connected>\n");
+    }
+
+    return status;
+}
+
+static ucs_status_t ucp_ep_vfs_read_addr(ucp_ep_h ep, ucp_ep_vfs_attr_t *arg,
+                                         ucs_string_buffer_t *strb,
+                                         struct sockaddr *sockaddr)
+{
+    ucp_ep_attr_t attr;
+    ucs_status_t status;
+    struct sockaddr_storage *ss;
+
+    status = ucp_ep_vfs_query_sockaddr(ep, &attr, arg->field_mask, strb);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ss = (struct sockaddr_storage*)UCS_PTR_BYTE_OFFSET(&attr, arg->offset);
+
+    return ucs_sockaddr_copy(sockaddr, (struct sockaddr*)ss);
+}
+
+static void ucp_ep_vfs_read_addr_ip(void *obj, ucs_string_buffer_t *strb,
+                                    void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
+
+    if (ucp_ep_vfs_read_addr((ucp_ep_h)obj, (ucp_ep_vfs_attr_t*)arg_ptr, strb,
+                             &sockaddr) == UCS_OK) {
+        ucp_vfs_read_ip(&sockaddr, strb);
+    }
+}
+
+static void ucp_ep_vfs_read_addr_port(void *obj, ucs_string_buffer_t *strb,
+                                      void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
+
+    if (ucp_ep_vfs_read_addr((ucp_ep_h)obj, (ucp_ep_vfs_attr_t*)arg_ptr, strb,
+                             &sockaddr) == UCS_OK) {
+        ucp_vfs_read_port(&sockaddr, strb);
+    }
+}
+
+static void ucp_ep_vfs_init_address(ucp_ep_h ep)
+{
+    ucp_ep_attr_t attr;
+    struct sockaddr *sockaddr;
+    size_t i;
+    ucp_wireup_ep_t *cm_wireup_ep;
+
+    for (i = 0; i < ucs_static_array_size(ucp_ep_vfs_attrs); ++i) {
+        attr.field_mask = ucp_ep_vfs_attrs[i].field_mask;
+        if (ucp_ep_query_sockaddr(ep, &attr) == UCS_OK) {
+            sockaddr = (struct sockaddr*)
+                    UCS_PTR_BYTE_OFFSET(&attr, ucp_ep_vfs_attrs[i].offset);
+        } else if ((cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ep)) != NULL) {
+            sockaddr = (struct sockaddr*)&cm_wireup_ep->cm_remote_sockaddr;
+        } else {
```

Comment:
seems this logic should be part of ucp_ep_query_sockaddr

Reply:
Changing `ucp_ep_query_sockaddr` will change the implementation of UCP interface `ucp_ep_query`.
Do we really want to change the logic of existing implementation?

Reply:
yes: it better return sockaddr if have it, even if not fully connected

Reply:
Removed check for wire up endpoint here, because the wire ep is used by `ucp_ep_get_cm_uct_ep` which retrieves the UCT endpoint. The UCT endpoint, in turn, is used by `ucp_ep_query_sockaddr`.